core/src/identity/rsa.rs

```diff
@@ -123,93 +123,148 @@ impl fmt::Debug for PublicKey {
-        DerObject {
-            tag: DerTag::x06,
-            value: DerValue {
-                data: vec![ 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01 ]
-            }
-        };
-}
+const ASN1_OBJECT_IDENTIFIER_TAG: u8 = 6;
+/// The DER encoding of the object identifier (OID) 'rsaEncryption' for
+/// RSA public keys defined for X.509 in [RFC-3279] and used in
+/// SubjectPublicKeyInfo structures defined in [RFC-5280].
+///
+/// [RFC-3279]: https://tools.ietf.org/html/rfc3279#section-2.3.1
+/// [RFC-5280]: https://tools.ietf.org/html/rfc5280#section-4.1
+const ASN1_RSA_ENCRYPTION_OID: [u8;9] = [ 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01 ];
 
 /// The ASN.1 OID for "rsaEncryption".
 #[derive(Clone)]
 struct Asn1OidRsaEncryption();
 
-impl IntoDerObject for Asn1OidRsaEncryption {
-    fn into_der_object(self) -> DerObject {
-        OID_RSA_ENCRYPTION_DER.clone()
-    }
-    fn serialized_len(&self) -> usize {
-        OID_RSA_ENCRYPTION_DER.serialized_len()
+impl asn1_der::typed::DerEncodable for Asn1OidRsaEncryption {
+    fn encode<S: asn1_der::Sink>(&self, sink: &mut S) -> Result<(), asn1_der::Asn1DerError> {
+        // TODO: `DerObject::write` is a hidden method. Should one use `new` or `new_from_source`?
+        // If so, I don't see a way to do that given that `S: Into<&'a [u8]` would be required.
+        asn1_der::DerObject::write(
+            ASN1_OBJECT_IDENTIFIER_TAG,
+            ASN1_RSA_ENCRYPTION_OID.len(),
+            &mut ASN1_RSA_ENCRYPTION_OID.iter(),
+            sink,
+        )?;
+        Ok(())
     }
 }
 
-impl FromDerObject for Asn1OidRsaEncryption {
-    fn from_der_object(o: DerObject) -> Result<Self, Asn1DerError> {
-        if o.tag != DerTag::x06 {
-            return Err(Asn1DerError::InvalidTag)
+impl DerDecodable<'_> for Asn1OidRsaEncryption {
+    fn load(object: DerObject<'_>) -> Result<Self, Asn1DerError> {
+        if object.tag() != ASN1_OBJECT_IDENTIFIER_TAG {
+            return Err(Asn1DerError::new(
+                Asn1DerErrorVariant::InvalidData("DER object tag is not the object identifier tag."),
+            ));
         }
-        if o.value != OID_RSA_ENCRYPTION_DER.value {
-            return Err(Asn1DerError::InvalidEncoding)
+
+        if object.value() != ASN1_RSA_ENCRYPTION_OID {
+            return Err(Asn1DerError::new(
+                Asn1DerErrorVariant::InvalidData("DER object is not the 'rsaEncryption' identifier."),
+            ));
         }
+
         Ok(Asn1OidRsaEncryption())
     }
 }
 
 /// The ASN.1 AlgorithmIdentifier for "rsaEncryption".
-#[derive(Asn1Der)]
 struct Asn1RsaEncryption {
     algorithm: Asn1OidRsaEncryption,
     parameters: ()
 }
 
+impl asn1_der::typed::DerEncodable for Asn1RsaEncryption {
+    fn encode<S: asn1_der::Sink>(&self, sink: &mut S) -> Result<(), asn1_der::Asn1DerError> {
+        // TODO: Is there a way to write to `sink` directly?
+        let mut algorithm = Vec::new();
+        self.algorithm.encode(&mut algorithm)?;
```

Comment:
Is this the way to go?

Reply:
Well in general yes; however I've added an additional API to directly create a `DerObject` from an encodable object:

```rust
/// The ASN.1 AlgorithmIdentifier for "rsaEncryption".
struct Asn1RsaEncryption {
    algorithm: Asn1OidRsaEncryption,
    parameters: ()
}
impl<'a> DerDecodable<'a> for Asn1RsaEncryption {
    fn load(object: DerObject<'a>) -> Result<Self, Asn1DerError> {
        let sequence = Sequence::load(object)?;

        Ok(Self {
            algorithm: sequence.get_as(0)?,
            parameters: sequence.get_as(1)?
        })
    }
}
impl DerEncodable for Asn1RsaEncryption {
    fn encode<S: Sink>(&self, sink: &mut S) -> Result<(), Asn1DerError> {
        // I've published an update to allow this; take a look at 0.7.4
        // The buffers are still necessary because a DerObject is a simple view
        let mut algorithm_buf = Vec::new();
        let algorithm = self.algorithm.der_object(VecBacking(&mut algorithm_buf))?;

        let mut parameters_buf = Vec::new();
        let parameters = self.parameters.der_object(VecBacking(&mut parameters_buf))?;
        
        // Create a sequence
        Sequence::write(&[algorithm, parameters], sink)
    }
}
```

I agree that this is not very straight elegant; the API was mainly designed with `serde_asn1_der` in mind. However if you're interested I can also revive `asn1_der_derive`, if `serde` is no option for you?